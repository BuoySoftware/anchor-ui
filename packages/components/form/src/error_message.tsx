import { WarningCircle } from "iconoir-react";
import { Box, BoxProps, Flex } from "@buoysoftware/anchor-layout";
import { Body, Markdown } from "@buoysoftware/anchor-typography";
import { useTheme } from "@buoysoftware/anchor-theme";

interface OwnProps {
  errors: string[];
  parentId: string;
}

type ErrorMessageProps = OwnProps & BoxProps;

export const ErrorMessage: React.FC<ErrorMessageProps> = ({
  errors,
  parentId,
  ...props
}): React.ReactElement => {
  const { colors } = useTheme();
  const id = `error-${parentId}`;

  return (
    <Flex
      bg="background.criticalSubdued"
      borderRadius="4px"
      data-testid={id}
      id={id}
      py="xxs"
      px="xs"
      role="alert"
      {...props}
    >
      <WarningCircle color={colors.text.critical} width={16} height={16} />
      <Box ml="xs">
        {errors.map((error, index) => (
          <Body as="div" size="m" key={`error-${index}`} color="text.critical">
            <Markdown>{error}</Markdown>
          </Body>
        ))}
      </Box>
    </Flex>
  );
};
