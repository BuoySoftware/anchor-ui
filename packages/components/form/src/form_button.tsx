import { useFormState } from "react-hook-form";
import { Button, ButtonProps } from "@buoysoftware/anchor-button";

type FormButtonProps = ButtonProps;

export const FormButton: React.FC<FormButtonProps> = ({
  children,
  ...props
}): React.ReactElement => {
  const { isSubmitting } = useFormState();

  return (
    <Button submitting={isSubmitting} {...props}>
      {children}
    </Button>
  );
};
