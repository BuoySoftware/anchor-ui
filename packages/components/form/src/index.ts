export * from "./checkbox";
export * from "./checkbox_field";
export * from "./error_message";
export * from "./form";
export * from "./form_base_errors";
export * from "./form_button";
export * from "./form_scope_provider";
export * from "./input";
export * from "./label";
export * from "./radio_field";
export * from "./select";
export * from "./select_field";
export * from "./text_area";
export * from "./text_field";
export * from "./use_conditional_field";
export * from "./use_form_field";

export { default as anchorFormsEn } from "./i18n/anchorForms.en.json";
