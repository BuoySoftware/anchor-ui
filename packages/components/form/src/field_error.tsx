import compact from "lodash/compact";
import flatten from "lodash/flatten";
import { FieldError as HookFormFieldError } from "react-hook-form";
import snakeCase from "lodash/snakeCase";
import { BoxProps } from "@buoysoftware/anchor-layout";
import { TFunction } from "i18next";
import { useTranslation } from "react-i18next";

import { useFormScope } from "./form_scope_provider";
import { ErrorMessage } from "./error_message";

export type InputType =
  | "checkbox"
  | "collectionCheckbox"
  | "hidden"
  | "phoneNumber"
  | "radio"
  | "select"
  | "text";

interface OwnProps {
  error?: HookFormFieldError;
  name?: string;
  inputId: string;
  inputType?: InputType;
}

export type FieldErrorProps = OwnProps & BoxProps;

/**
 * Look up an error message in the i18next translations. This utilizes the
 * i18next fallback mechanism for determining who to translate the error message.
 *
 * The order is as follows:
 *
 * 1. An error message in a custom provided namespace via tNamespace scoped to
 *    forms.errors and then the form scope, field name, and error typea.
 *
 * 2. An error message in the default forms namespace scoped to the form scope,
 *    field name and error type.
 *
 * 3. An error message in default forms namespace for the given input type and
 *    error type.
 *
 * 4. An error message in the anchorForms namespace for the given input type and
 *    error type.
 *
 * 5. An error message in the forms namespace for the given error type
 *
 * 6. An error message in the anchorForms namespace for the given error type
 */
function lookupErrorMessage(
  error: HookFormFieldError,
  scope: string,
  name: string,
  inputType: InputType,
  t: TFunction,
  tNamespace: string[]
): string {
  if (error.message !== undefined && error.message !== "") {
    return error.message;
  }

  const i18nName = snakeCase(name);
  const i18nErrorScope = snakeCase(scope);

  // These are translations scoped to the form's scope and field name.
  // The use of forms.errors is to look into the custom tNamespace if one was
  // provided. The second will look up in the forms default namespace.
  const fieldScopes = [
    `forms.errors.${i18nErrorScope}.${i18nName}.${error.type}`,
    `errors.${i18nErrorScope}.${i18nName}.${error.type}`,
  ];

  // Global scopes are used to look up defaults in the forms namespace or
  // the anchorForms namespace
  const globalScopes = [
    `errors.input_types.${snakeCase(inputType)}.${error.type}`,
    `errors.${error.type}`,
  ];

  const combinedGlobalScopes = tNamespace.flatMap((namespace: string) =>
    globalScopes.map((scope) => `${namespace}:${scope}`)
  );

  return t([...fieldScopes, ...combinedGlobalScopes]);
}

export const FieldError: React.FC<FieldErrorProps> = ({
  error,
  inputId,
  name,
  inputType = "text",
  ...elementProps
}): React.ReactElement | null => {
  const { scope, tNamespace = [] } = useFormScope();
  const namespace = compact(flatten([tNamespace, "forms", "anchorForms"]));
  const { t } = useTranslation(namespace);

  if (!error || !name) return null;

  const message = lookupErrorMessage(
    error,
    scope,
    name,
    inputType,
    t,
    namespace
  );

  return (
    <ErrorMessage
      errors={message.split("\n")}
      parentId={inputId}
      {...elementProps}
    />
  );
};
