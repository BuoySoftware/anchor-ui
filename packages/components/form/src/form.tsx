import kebabCase from "lodash/kebabCase";
import { useEffect, useState } from "react";
import {
  FieldValues,
  FormProvider,
  get,
  SubmitHandler,
  UseFormReturn,
} from "react-hook-form";
import { ErrorMessage } from "./error_message";
import { Box, BoxProps } from "@buoysoftware/anchor-layout";

import { FormScopeProvider, TNamespace } from "./form_scope_provider";

export type FormValidationErrors<T> = {
  [key in keyof T]: string[];
};

export type ErrorKeys<TFormValues> = Record<string, keyof TFormValues>;

interface OwnProps<TFormValues extends FieldValues> {
  "data-testid"?: string;
  id: string;
  scope: string;
  onSubmit?: SubmitHandler<TFormValues>;
  children: React.ReactNode;
  form: UseFormReturn<TFormValues>;
  errorKeys?: ErrorKeys<TFormValues>;
  mutationErrors?: Partial<FormValidationErrors<TFormValues>>;
  tNamespace?: TNamespace;
}

export type FormProps<TFormValues extends FieldValues> = OwnProps<TFormValues> &
  BoxProps;

export const Form = <TFormValues extends FieldValues>({
  "data-testid": testId,
  id,
  scope,
  children,
  errorKeys = {},
  form: { handleSubmit, ...form },
  onSubmit = () => undefined,
  mutationErrors = {},
  tNamespace,
  ...elementProps
}: FormProps<TFormValues>): React.ReactElement => {
  const [hasErrors, setHasErrors] = useState(false);

  useEffect(() => {
    if (hasErrors) {
      Object.entries(mutationErrors).forEach(([key, errors]) => {
        if (key !== "base") {
          const messages = errors?.join("\n");
          const inputKey = get(errorKeys, key, key);

          form.setError(inputKey as any, {
            type: "server",
            message: messages,
          });
        }
      });
    }
  }, [hasErrors, mutationErrors]);

  useEffect(() => {
    setHasErrors(!!Object.keys(mutationErrors).length);
  }, [mutationErrors]);

  const baseErrors = mutationErrors["base"] || [];

  return (
    <FormScopeProvider value={{ formId: id, baseErrors, scope, tNamespace }}>
      <FormProvider {...form} handleSubmit={handleSubmit}>
        <Box
          as="form"
          autoComplete="off"
          id={id}
          data-testid={testId || kebabCase(scope)}
          onSubmit={handleSubmit(onSubmit)}
          {...elementProps}
        >
          {children}
        </Box>
      </FormProvider>
    </FormScopeProvider>
  );
};
