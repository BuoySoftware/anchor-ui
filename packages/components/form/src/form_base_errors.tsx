import { useFormScope } from "./form_scope_provider";
import { ErrorMessage } from "./error_message";

export const FormBaseErrors: React.FC = (): React.ReactElement | null => {
  const { baseErrors, formId } = useFormScope();

  if (baseErrors.length === 0) return null;

  return (
    <ErrorMessage errors={baseErrors} parentId={formId} mb="form.fieldGap" />
  );
}
