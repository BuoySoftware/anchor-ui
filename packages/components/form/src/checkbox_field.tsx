import { Box, BoxProps } from "@buoysoftware/anchor-layout";
import { RegisterOptions, useFormContext } from "react-hook-form";

import { useFormField } from "./use_form_field";
import { InlineLabel } from "./inline_label";
import { Checkbox } from "./checkbox";
import { FieldError } from "./field_error";

interface OwnProps {
  name: string;
  value?: string;
  rules?: RegisterOptions;
}

type CheckboxFieldProps = OwnProps & BoxProps;

export type CheckboxFieldValue = "1" | false;

export const DEFAULT_CHECKED_VALUE: CheckboxFieldValue = "1";

export const CheckboxField: React.FC<CheckboxFieldProps> = ({
  name,
  value = DEFAULT_CHECKED_VALUE,
  rules,
  ...boxProps
}): React.ReactElement => {
  const { label, error, inputId } = useFormField({ name });
  const { register } = useFormContext();

  return (
    <Box
      as="fieldset"
      data-testid={`checkbox-field-${inputId}`}
      display="flex"
      flexDirection="column"
      mb="form.fieldGap"
      {...boxProps}
    >
      <InlineLabel
        data-testid={`label-${inputId}`}
        display="flex"
        alignItems="center"
        htmlFor={inputId}
      >
        <Checkbox
          aria-describedby={error ? `error-${name}` : undefined}
          aria-invalid={error ? "true" : "false"}
          data-testid={inputId}
          id={inputId}
          mr="xxs"
          value={value}
          {...register(name, { shouldUnregister: true, ...rules })}
        />
        {label}
      </InlineLabel>
      <FieldError
        error={error}
        inputId={inputId}
        inputType="checkbox"
        mt="s"
        name={name}
      />
    </Box>
  );
};
