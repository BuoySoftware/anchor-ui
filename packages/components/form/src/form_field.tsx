import kebabCase from "lodash/kebabCase";
import styled from "styled-components";
import { FieldError as HookFormFieldError } from "react-hook-form";
import { Box, BoxProps } from "@buoysoftware/anchor-layout";
import { Label } from "./label";

import { FieldError, InputType } from "./field_error";

interface OwnProps {
  children: React.ReactNode;
  error?: HookFormFieldError;
  inputId: string;
  inputType?: InputType;
  label: string;
  name?: string;
}

export type FormFieldProps = OwnProps & Omit<BoxProps, "errors">;

export const FormField = styled(
  ({ children, error, inputId, inputType, name, label, ...elementProps }) => {
    const testId = kebabCase(inputId);

    return (
      <Box
        data-testid={`form-field-${testId}`}
        mb="form.fieldGap"
        display="flex"
        width="input"
        flexDirection="column"
        {...elementProps}
      >
        <Label htmlFor={inputId} mb="form.labelGap">
          {label}
        </Label>
        <Box position="relative" zIndex={1}>
          {children}
        </Box>
        <FieldError
          error={error}
          inputId={inputId}
          inputType={inputType}
          mt="xxs"
          width="input"
          name={name}
        />
      </Box>
    );
  }
)<FormFieldProps>``;
