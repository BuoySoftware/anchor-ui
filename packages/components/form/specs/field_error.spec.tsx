import i18n from "i18next";
import { initReactI18next } from "react-i18next";
import { render } from "@testing-library/react";

import { FormScopeProvider } from "../src";
import { FieldError } from "../src/field_error";

describe("<FieldError />", () => {
  const translations = {
    errors: {
      date: "Error type message",
      input_types: {
        text: {
          required: "Default text input required message",
        },
      },
      test_form: {
        my_field: {
          required: "Scoped required error",
        },
      },
    },
  };

  const customScopeTranslations = {
    forms: {
      errors: {
        test_form: {
          custom_field: {
            required: "Custom scoped required error",
          },
        },
      },
    },
  };

  i18n.use(initReactI18next).init({
    lng: "en",
    ns: ["forms", "customScope"],
    resources: {
      en: {
        customScope: customScopeTranslations,
        forms: translations,
      },
    },
  });

  context("error is undefined", () => {
    it("renders nothing", () => {
      const errors = render(
        <FieldError name="my-field" inputId="test-input" />
      );

      expect(errors.asFragment()).toMatchSnapshot();
    });
  });

  context("error field name is missing", () => {
    it("renders nothing", () => {
      const errors = render(
        <FieldError
          error={{ type: "required", message: "Is required" }}
          inputId="test-input"
        />
      );

      expect(errors.asFragment()).toMatchSnapshot();
    });
  });

  context("error message is provided", () => {
    context("with a single message", () => {
      it("renders the provided message", () => {
        const errors = render(
          <FieldError
            name="myField"
            error={{ type: "server", message: "Invalid field" }}
            inputId="test-input"
          />
        );

        expect(errors.asFragment()).toMatchSnapshot();
      });
    });

    context("with multiple messages", () => {
      it("renders all of the provided messages", () => {
        const errors = render(
          <FieldError
            name="myField"
            error={{ type: "server", message: "Invalid field\nOther error" }}
            inputId="test-input"
          />
        );

        expect(errors.asFragment()).toMatchSnapshot();
      });
    });
  });

  context("error message is not provided", () => {
    context("scope and field error exists", () => {
      it("renders a message tailored to the scope and field name", () => {
        const errors = render(
          <FormScopeProvider value={{ scope: "testForm" }}>
            <FieldError
              name="myField"
              error={{ type: "required", message: "" }}
              inputId="test-input"
            />
          </FormScopeProvider>
        );

        expect(errors.getByTestId("error-test-input")).toHaveTextContent(
          "Scoped required error"
        );
      });

      it("supports a custom tNamespace for the scope and field error", () => {
        const errors = render(
          <FormScopeProvider
            value={{ scope: "testForm", tNamespace: ["customScope"] }}
          >
            <FieldError
              name="customField"
              error={{ type: "required", message: "" }}
              inputId="test-input"
            />
          </FormScopeProvider>
        );

        expect(errors.getByTestId("error-test-input")).toHaveTextContent(
          "Custom scoped required error"
        );
      });
    });

    context("input type message exists", () => {
      it("renders the input type validation error message", () => {
        const errors = render(
          <FormScopeProvider value={{ scope: "testForm" }}>
            <FieldError
              name="testField2"
              inputType="text"
              error={{ type: "required", message: "" }}
              inputId="test-input"
            />
          </FormScopeProvider>
        );

        expect(errors.getByTestId("error-test-input")).toHaveTextContent(
          "Default text input required message"
        );
      });
    });

    it("falls back to a message based off the validation error type", () => {
      const errors = render(
        <FormScopeProvider value={{ scope: "testForm" }}>
          <FieldError
            name="date"
            inputType="text"
            error={{ type: "date", message: "" }}
            inputId="test-input"
          />
        </FormScopeProvider>
      );

      expect(errors.getByTestId("error-test-input")).toHaveTextContent(
        "Error type message"
      );
    });
  });

  context("error message with markdown", () => {
    context("with a single message", () => {
      it("renders the provided message with markup", () => {
        const errors = render(
          <FieldError
            name="myField"
            error={{
              type: "server",
              message: "This is **an error** with markdown",
            }}
            inputId="test-input"
          />
        );

        expect(errors.asFragment()).toMatchSnapshot();
      });
    });
  });
});
