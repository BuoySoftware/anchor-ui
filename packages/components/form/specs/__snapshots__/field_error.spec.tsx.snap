// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldError /> error field name is missing renders nothing 1`] = `<DocumentFragment />`;

exports[`<FieldError /> error is undefined renders nothing 1`] = `<DocumentFragment />`;

exports[`<FieldError /> error message is provided with a single message renders the provided message 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe jZDVhQ"
    data-testid="error-test-input"
    id="error-test-input"
    role="alert"
  >
    <div
      class="sc-bcXHqe sc-hLBbgP sc-eDvSVe bGCHBu rKWEz"
      color="text.critical"
      font-size="font-size-100"
      font-style="none"
      font-weight="400"
    >
      <p
        class="sc-bcXHqe sc-hLBbgP dYCYGe dDZkPR"
        color="inherit"
      >
        Invalid field
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<FieldError /> error message is provided with multiple messages renders all of the provided messages 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe jZDVhQ"
    data-testid="error-test-input"
    id="error-test-input"
    role="alert"
  >
    <div
      class="sc-bcXHqe sc-hLBbgP sc-eDvSVe bGCHBu rKWEz"
      color="text.critical"
      font-size="font-size-100"
      font-style="none"
      font-weight="400"
    >
      <p
        class="sc-bcXHqe sc-hLBbgP dYCYGe dDZkPR"
        color="inherit"
      >
        Invalid field
      </p>
    </div>
    <div
      class="sc-bcXHqe sc-hLBbgP sc-eDvSVe bGCHBu rKWEz"
      color="text.critical"
      font-size="font-size-100"
      font-style="none"
      font-weight="400"
    >
      <p
        class="sc-bcXHqe sc-hLBbgP dYCYGe dDZkPR"
        color="inherit"
      >
        Other error
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<FieldError /> error message with markdown with a single message renders the provided message with markup 1`] = `
<DocumentFragment>
  <div
    class="sc-bcXHqe jZDVhQ"
    data-testid="error-test-input"
    id="error-test-input"
    role="alert"
  >
    <div
      class="sc-bcXHqe sc-hLBbgP sc-eDvSVe bGCHBu rKWEz"
      color="text.critical"
      font-size="font-size-100"
      font-style="none"
      font-weight="400"
    >
      <p
        class="sc-bcXHqe sc-hLBbgP dYCYGe dDZkPR"
        color="inherit"
      >
        This is 
        <strong
          class="sc-bcXHqe sc-hLBbgP dYCYGe ePgUZj"
          color="inherit"
          font-weight="bold"
        >
          an error
        </strong>
         with markdown
      </p>
    </div>
  </div>
</DocumentFragment>
`;
